gax/src/main/java/com/google/api/gax/rpc/mtls/MtlsProvider.java

```diff
@@ -0,0 +1,197 @@
+   * mTLS endpoint; AUTO means auto switch to mTLS endpoint if client certificate exists and should
+   * be used.
+   */
+  public UseMtlsEndpoint useMtlsEndpoint() {
+    String useMtlsEndpoint = envProvider.getenv("GOOGLE_API_USE_MTLS_ENDPOINT");
+    if ("never".equals(useMtlsEndpoint)) {
+      return UseMtlsEndpoint.NEVER;
+    } else if ("always".equals(useMtlsEndpoint)) {
+      return UseMtlsEndpoint.ALWAYS;
+    }
+    return UseMtlsEndpoint.AUTO;
+  }
+
+  /** The mutual TLS key store created with the default client certificate on device. */
+  public KeyStore getKeyStore() throws IOException {
+    try (InputStream stream = new FileInputStream(metadataPath)) {
+      return getKeyStore(stream, processProvider);
+    } catch (FileNotFoundException ignored) {
+      // file doesn't exist
+      return null;
+    }
+  }
+
+  @VisibleForTesting
+  static KeyStore getKeyStore(InputStream metadata, ProcessProvider processProvider)
+      throws IOException {
+    try {
+      Process process = processProvider.createProcess(metadata);
+
+      // Run the command and timeout after 1000 milliseconds.
+      int exitCode = runCertificateProviderCommand(process, 1000);
+      if (exitCode != 0) {
+        throw new IOException("Cert provider command failed with exit code: " + exitCode);
+      }
+
+      // Create mTLS key store with the input certificates from shell command.
+      return SecurityUtils.createMtlsKeyStore(process.getInputStream());
+    } catch (InterruptedException e) {
+      throw new IOException("Interrupted executing certificate provider command", e);
+    } catch (GeneralSecurityException e) {
+      throw new IOException("Failed to get key store", e);
+    }
+  }
+
+  @VisibleForTesting
+  static List<String> extractCertificateProviderCommand(InputStream contextAwareMetadata)
+      throws IOException {
+    JsonParser parser = new GsonFactory().createJsonParser(contextAwareMetadata);
+    ContextAwareMetadataJson json = parser.parse(ContextAwareMetadataJson.class);
+    return json.getCommands();
+  }
+
+  @VisibleForTesting
+  static int runCertificateProviderCommand(Process commandProcess, long timeoutMilliseconds)
+      throws IOException, InterruptedException {
+    long startTime = System.currentTimeMillis();
+    long remainTime = timeoutMilliseconds;
+    boolean terminated = false;
+
+    while (remainTime > 0) {
+      Thread.sleep(Math.min(remainTime + 1, 100));
+      remainTime -= System.currentTimeMillis() - startTime;
+
+      try {
+        // Check if process is terminated by polling the exitValue, which throws
+        // IllegalThreadStateException if not terminated.
+        commandProcess.exitValue();
+        terminated = true;
+        break;
+      } catch (IllegalThreadStateException ignored) {
+      }
+    }
+
+    if (!terminated) {
+      commandProcess.destroy();
+      throw new IOException("cert provider command timed out");
+    }
+
+    return commandProcess.exitValue();
```

Comment:
could this change between line 183 and here?

Reply:
I don't see any problems with this part. Any concerns?